velox/functions/sparksql/aggregates/ModeAggregate.cpp

```diff
@@ -0,0 +1,582 @@
+      values.~TValues();
+    }
+
+   private:
+    // Stores the non-inlined string in memory blocks managed by
+    // HashStringAllocator, if it's not exist in the values map.
+    StringView store(StringView value, HashStringAllocator* allocator) {
+      if (!value.isInline()) {
+        auto it = values.find(value);
+        if (it == values.end()) {
+          value = strings.append(value, *allocator);
+        }
+      }
+      return value;
+    }
+  };
+};
+
+struct ComplexTypeAccumulator {
+  using ValueMap = folly::F14FastMap<
+      AddressableNonNullValueList::Entry,
+      int64_t,
+      AddressableNonNullValueList::Hash,
+      AddressableNonNullValueList::EqualTo,
+      AlignedStlAllocator<
+          std::pair<const AddressableNonNullValueList::Entry, int64_t>,
+          16>>;
+
+  ValueMap values;
+
+  AddressableNonNullValueList serializedValues;
+
+  ComplexTypeAccumulator(const TypePtr& type, HashStringAllocator* allocator)
+      : values{
+            0,
+            AddressableNonNullValueList::Hash{},
+            AddressableNonNullValueList::EqualTo{type},
+            AlignedStlAllocator<
+                std::pair<const AddressableNonNullValueList::Entry, int64_t>,
+                16>(allocator)} {}
+
+  size_t size() const {
+    return values.size();
+  }
+
+  void addValue(
+      DecodedVector& decoded,
+      vector_size_t index,
+      HashStringAllocator* allocator) {
+    addValueWithCount(*decoded.base(), decoded.index(index), 1, allocator);
+  }
+
+  void addValueWithCount(
+      const BaseVector& keys,
+      vector_size_t index,
+      int64_t count,
+      HashStringAllocator* allocator) {
+    auto entry = serializedValues.append(keys, index, allocator);
+
+    auto it = values.find(entry);
+    if (it == values.end()) {
+      // New entry.
+      values[entry] += count;
+    } else {
+      // Existing entry.
+      serializedValues.removeLast(entry);
+
+      it->second += count;
+    }
+  }
+
+  void extractValues(
+      BaseVector& keys,
+      FlatVector<int64_t>& counts,
+      vector_size_t offset) {
+    auto index = offset;
+    for (const auto& [position, count] : values) {
+      AddressableNonNullValueList::read(position, keys, index);
+      counts.set(index, count);
+      ++index;
+    }
+  }
+
+  void extractValues(const VectorPtr& results, vector_size_t offset) {
+    auto iterator = std::max_element(
+        std::begin(values),
+        std::end(values),
+        [](const std::pair<AddressableNonNullValueList::Entry, int64_t>& p1,
+           const std::pair<AddressableNonNullValueList::Entry, int64_t>& p2) {
+          return p1.second < p2.second;
+        });
+    AddressableNonNullValueList::read(iterator->first, *results, offset);
+  }
+
+  void free(HashStringAllocator& allocator) {
+    serializedValues.free(allocator);
+    using TValues = decltype(values);
+    values.~TValues();
+  }
+};
+
+template <typename T>
+class ComplexTypeModeAggregate : public Aggregate {
```

Comment:
Could you clarify the reason that simple-function interface cannot be used?

Reply:
`ComplexTypeAccumulator` needs inputType parameter, which doesn't align with `SimpleAggregateAdapter`'s intializer's signature.
https://github.com/facebookincubator/velox/blob/84428f9851ec0f0a04b5e75160b6400549366f8e/velox/exec/SimpleAggregateAdapter.h#L44C12-L44C34

Reply:
I see that [HistogramAggregate.cpp](https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/aggregates/HistogramAggregate.cpp) adds the inputType parameter for all accumulators and does not use it for non-complex types, In this way the Aggregate class could be reused for all types. Which one is better?

Reply:
Yes, it's not not used for scalar types.
https://github.com/facebookincubator/velox/blob/84428f9851ec0f0a04b5e75160b6400549366f8e/velox/functions/prestosql/aggregates/HistogramAggregate.cpp#L44

I actaully tried to implement mode agg like Presto HistogramAggregate and compared it with the implementation now.
In that way we still needs to implement ScalarAccumulator and StringViewAccumulator, which should be very similar with ModeAggregate and StringModeAggregate. Besides we need to add if-else logic to handle the type difference.
https://github.com/facebookincubator/velox/blob/84428f9851ec0f0a04b5e75160b6400549366f8e/velox/functions/prestosql/aggregates/HistogramAggregate.cpp#L256
https://github.com/facebookincubator/velox/blob/84428f9851ec0f0a04b5e75160b6400549366f8e/velox/functions/prestosql/aggregates/HistogramAggregate.cpp#L306

This is the reason I prefer the implementation now, please let mw know your point, thanks.

Reply:
Kindly ping, thanks